src/libsyntax/test.rs

```diff
@@ -639,184 +398,49 @@ fn path_name_i(idents: &[Ident]) -> String {
-
-    // Rather than directly give the test function to the test
-    // harness, we create a wrapper like one of the following:
-    //
-    //     || test::assert_test_result(real_function()) // for test
-    //     |b| test::assert_test_result(real_function(b)) // for bench
-    //
-    // this will coerce into a fn pointer that is specialized to the
-    // actual return type of `real_function` (Typically `()`, but not always).
-    let fn_expr = {
-        // construct `real_function()` (this will be inserted into the overall expr)
-        let real_function_expr = ecx.expr_path(ecx.path_global(span, visible_path));
-        // construct path `test::assert_test_result`
-        let assert_test_result = test_path("assert_test_result");
-        if test.bench {
-            // construct `|b| {..}`
-            let b_ident = Ident::with_empty_ctxt(Symbol::gensym("b"));
-            let b_expr = ecx.expr_ident(span, b_ident);
-            ecx.lambda(
-                span,
-                vec![b_ident],
-                // construct `assert_test_result(..)`
-                ecx.expr_call(
-                    span,
-                    ecx.expr_path(assert_test_result),
-                    vec![
-                        // construct `real_function(b)`
-                        ecx.expr_call(
-                            span,
-                            real_function_expr,
-                            vec![b_expr],
-                        )
-                    ],
-                ),
-            )
-        } else {
-            // construct `|| {..}`
-            ecx.lambda(
-                span,
-                vec![],
-                // construct `assert_test_result(..)`
-                ecx.expr_call(
-                    span,
-                    ecx.expr_path(assert_test_result),
-                    vec![
-                        // construct `real_function()`
-                        ecx.expr_call(
-                            span,
-                            real_function_expr,
-                            vec![],
-                        )
-                    ],
-                ),
-            )
-        }
-    };
-
-    let variant_name = if test.bench { "StaticBenchFn" } else { "StaticTestFn" };
+    }
+    visible_path.extend_from_slice(path);
+    visible_path
+}
 
-    // self::test::$variant_name($fn_expr)
-    let testfn_expr = ecx.expr_call(span, ecx.expr_path(test_path(variant_name)), vec![fn_expr]);
+fn is_test_case(i: &ast::Item) -> bool {
+    attr::contains_name(&i.attrs, "test_case")
+}
 
-    // self::test::TestDescAndFn { ... }
-    ecx.expr_struct(span,
-                    test_path("TestDescAndFn"),
-                    vec![field("desc", desc_expr),
-                         field("testfn", testfn_expr)])
+fn get_test_runner(sd: &errors::Handler, krate: &ast::Crate) -> Option<ast::Path> {
+    if let Some(test_attr) = attr::find_by_name(&krate.attrs, "test_runner") {
```

Comment:
Could this use `?` and early returns on invalid cases to reduce the indentation a bit?